common/remote/hooks/s3/s3file.cpp

```diff
@@ -16,700 +16,1320 @@
-        //ever calls close, and it would be harmless (and would complicate the rest of the code).
+        return (planeName == other.planeName) && (device == other.device);
     }
+};
 
-    // Write methods not implemented - this is a read-only file
-    virtual size32_t write(offset_t pos, size32_t len, const void * data) override
+// Hash specialization for S3ClientKey
+namespace std {
+    template<>
+    struct hash<S3ClientKey>
     {
-        throwUnexpected();
-        return 0;
-    }
-    virtual void setSize(offset_t size) override
+        size_t operator()(const S3ClientKey& key) const noexcept
+        {
+            unsigned h = 0;
+            if (key.planeName.str())
+                h = hashc((const unsigned char*)key.planeName.str(), key.planeName.length(), h);
+            h = hashvalue(key.device, h);
+            return h;
+        }
+    };
+}
+
+class S3ClientManager
+{
+private:
+    mutable CriticalSection cs;
+    std::unordered_map<S3ClientKey, std::unique_ptr<Aws::S3::S3Client>> clients;
+    bool initialized = false;
+
+public:
+    Aws::S3::S3Client& getClient(const char* planeName, unsigned device)
     {
-        throwUnexpected();
+        CriticalBlock block(cs);
+        if (!initialized)
+        {
+            initAWS();
+            initialized = true;
+        }
+
+        S3ClientKey key;
+        key.planeName.set(planeName);
+        key.device = device;
+
+        if (auto it = clients.find(key); it != clients.end())
+        {
+            return *(it->second);
+        }
+        else
+        {
+            std::unique_ptr<Aws::S3::S3Client> client = createClient(planeName, device);
+            Aws::S3::S3Client& ref = *client;
+            clients.emplace(key, std::move(client));
+            return ref;
+        }
     }
-    virtual void flush() override
+
+    void cleanup()
     {
-        //Could implement if we use the async version of the putObject call.
+        CriticalBlock block(cs);
+        if (initialized)
+        {
+            // Destroy clients before shutting down AWS SDK
+            clients.clear();
+            shutdownAWS();
+            initialized = false;
+        }
     }
-    virtual unsigned __int64 getStatistic(StatisticKind kind) override;
-    virtual IFile * queryFile() const override;
 
-protected:
-    size_t extractDataFromResult(size_t offset, size_t length, void * target);
+    ~S3ClientManager()
+    {
+        cleanup();
+    }
 
-protected:
-    Linked<S3File> file;
-    CriticalSection cs;
-    offset_t startResultOffset = 0;
-    offset_t endResultOffset = 0;
-    Aws::S3::Model::GetObjectOutcome readResult;
-    FileIOStats stats;
+private:
+    std::unique_ptr<Aws::S3::S3Client> createClient(const char* planeName, unsigned device)
+    {
+        // Load plane configuration
+        Owned<const IPropertyTree> plane = getStoragePlaneConfig(planeName, true);
```

Comment:
new code should use queryStoragePlaneByName(), but the interface may need extending to provide the information that is needed.  Better revisited as a separate PR.